File: app/src/main/java/com/expensivebelly/dagger2retaingraph/fragment/master/MasterPresenter.java
package com.expensivebelly.dagger2retaingraph.fragment.master;

import io.reactivex.Single;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

public class MasterPresenter {
    private final Single<Long> timeWaster;
    private MasterView view;
    private Disposable disposable;

    public MasterPresenter(Single<Long> timeWaster) {
        this.timeWaster = timeWaster;
    }

    void attachView(MasterView view) {
        this.view = view;
        view.init();
        startTimeWasting();
    }

    private void startTimeWasting() {
        disposable = timeWaster.subscribe(new Consumer<Long>() {
            @Override
            public void accept(Long aLong) throws Exception {
                view.hideProgressBar();
            }
        });
    }

    void detachView() {
        view = new NullMasterView();
        disposable.dispose();
    }
}

File: app/src/main/java/com/expensivebelly/dagger2retaingraph/fragment/master/MasterPresenter.kt
package com.expensivebelly.dagger2retaingraph.fragment.master

import com.expensivebelly.dagger2retaingraph.core.IPresenter
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.kotlin.plusAssign
import io.reactivex.rxjava3.kotlin.subscribeBy

class MasterPresenter(private val timeWaster: Single<Long>) : IPresenter<MasterView> {

    private val compositeDisposable = CompositeDisposable()

    override fun attach(view: MasterView) {
        compositeDisposable +=
                timeWaster.observeOn(AndroidSchedulers.mainThread())
                        .subscribeBy { view.hideProgressBar() }
    }

    override fun detach() {
        compositeDisposable.clear()
    }

}